File: linked_list/cxx/find_middle.cxx
#include "LinkedList.h"

ListNode* find_middle(ListNode* head)
{
	if(head == NULL)
		return head;

	ListNode *slow = head, *fast = head;
	while(fast->next and fast->next->next)
	{
		slow = slow->next;
		fast = fast->next->next;
	}
	return slow;
}

File: linked_list/py3/find_middle.py
# For even length
# O O $
#sf
# ^ s f
#
# For odd length
# O O O $
#sf
# ^ s f
# ^   s   f
# Time:  O(n)
# Space: O(1)

# For even length
# O O $
# s f
#
# For odd length
# O O O $
# s f
#   s   f
# Time:  O(n)
# Space: O(1)

def find_middle(head):
    fast = slow = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    return slow
